Initial support for Qualcomm IPQ5424 platform

quic-varada opened the pull request:
This series aims to enable the initial support for the Qualcomm IPQ5424 based RDP466 boards. Please see [1] for SoC related details

Thanks
Varada

1 - https://docs.qualcomm.com/bundle/publicresource/87-82994-1_REV_B_Qualcomm_Networking_Pro_A7_Elite_Platform_Product_Brief.pdf

Commit:
qualcommbe: ipq95xx: Update PCS & PPE & QCA8084 drivers

Update the PCS & PPE patches from the upstream patches, which
enhance performance with GRO/GSO/TSO issue fixed.

Update the QCA8084 patches to support QCA8084 PHY function, which
can be used with IPQ9574 RDP connected with QCA8084.

Additionally, Add the RDP543 DTS to enable IPQ9574 RDP board connected
QCA8084 PHY chip.

Signed-off-by: Ram Chandra Jangir <rjangir@qti.qualcomm.com>

Commit:
qualcommbe: ipq54xx: Add initial support for RDP466 Ethernet function

IPQ54XX RDP466 is connected with two 10G QCA8111 PHYs and QCA8386
switch for the Ethernet function. Add the related drivers to enable
these Ethernet hardware blocks.

Signed-off-by: Luo Jie <quic_luoj@quicinc.com>
Signed-off-by: Ram Chandra Jangir <rjangir@qti.qualcomm.com>

hauke commented:
Please base this on top of #18796 which upgrades the target to kernel 6.12.

mrnuke commented on target/linux/qualcommbe/patches-6.6/901-05-net-phy-qca808x-Add-QCA8084-SerDes-probe-and-remove-.patch:
```diff
@@ -0,0 +1,448 @@
+ 	if (ret && ret != -EINVAL)
+ 		return ret;
+ 
++	for_each_available_child_of_node(np, child) {
++		struct mdio_device *mdiodev;
++
++		if (of_node_name_eq(child, "pcs-phy")) {
++			mdiodev = qca8084_package_pcs_probe(child);
++			if (IS_ERR(mdiodev))
++				return PTR_ERR(mdiodev);
++
++			shared_priv->mdiodev[0] = mdiodev;
++		}
++
++		if (of_node_name_eq(child, "xpcs-phy")) {
++			mdiodev = qca8084_package_xpcs_probe(child);
++			if (IS_ERR(mdiodev))
++				return PTR_ERR(mdiodev);
++
++			shared_priv->mdiodev[1] = mdiodev;
++		}
++	}
```
I don't think this is going to fly upstream. If **pcs-phy** and **xpcs-phy** children are available, then you will populate your data structures, and, if needed, elegantly abort with an error code. Perfect!

However, if either of the two children are unavailable, you'll have NULL pointers in `shared_priv->mdiodev[]`. You'll get a NULL pointer dereference and crash later down the road.

Suggest using `of_get_child_by_name()` instead.

jie-luo89 replied:
Thank you for your suggestions. This code will be improved as recommended. Both the "pcs-phy" and "xpcs-phy" nodes should be available for the QCA8084 PHY mode, and the macro `IS_ERR_OR_NULL()` will be used to handle NULL pointer checks appropriately.

mrnuke reviewed (commented):
Hi there! I am working on getting the qualcommbe target on linux-6.12. I am using the git branch at [1] to generate my patches. Do you also have a kernel git branch that was used to generate these patches?

I think there's some opportunity to integrate some of your changes, as they will also be useful for ipq95xx.

[1] https://github.com/mrnuke/linux/commits/ipq95xx-devel/

mrnuke commented on target/linux/qualcommbe/patches-6.6/901-03-arm64-dts-ipq9574-Update-the-IPQ9574-RDP433-DTS.patch:
```diff
@@ -0,0 +1,507 @@
++        status = "okay";
++};
++
++&pcs0_ch0 {
++        clocks = <&nsscc NSS_CC_UNIPHY_PORT1_RX_CLK>,
++                 <&nsscc NSS_CC_UNIPHY_PORT1_TX_CLK>;
++        clock-names = "rx",
++                      "tx";
++        status = "okay";
++};
++
++&pcs0_ch1 {
++        clocks = <&nsscc NSS_CC_UNIPHY_PORT2_RX_CLK>,
++                 <&nsscc NSS_CC_UNIPHY_PORT2_TX_CLK>;
++        clock-names = "rx",
++                      "tx";
++        status = "okay";
++};
++
++&pcs0_ch2 {
++        clocks = <&nsscc NSS_CC_UNIPHY_PORT3_RX_CLK>,
++                 <&nsscc NSS_CC_UNIPHY_PORT3_TX_CLK>;
++        clock-names = "rx",
++                      "tx";
++        status = "okay";
++};
++
++&pcs0_ch3 {
++        clocks = <&nsscc NSS_CC_UNIPHY_PORT4_RX_CLK>,
++                 <&nsscc NSS_CC_UNIPHY_PORT4_TX_CLK>;
++        clock-names = "rx",
++                      "tx";
++        status = "okay";
++};
```
The PCS MIIs, PCS, and NSSCC clock controller are part of the SOC, right? So the clocks are routed internally in the SOC, and are not board-dependent, right?

So why are these moved from the SOC dtsi to the board DTS? Seems like every board now needs to duplicate the same PCS MII clock routing.

mrnuke replied:
Shouldn't then the `clocks` and `clock-names` properties be part of the DTSI?

jie-luo89 commented on target/linux/qualcommbe/patches-6.6/900-02-v6.15-net-phy-add-phylib-internal.h.patch:
```diff
@@ -0,0 +1,170 @@
+ drivers/net/phy/phy_device.c       |  2 ++
+ drivers/net/phy/phy_led_triggers.c |  2 ++
+ drivers/net/phy/phylib-internal.h  | 25 +++++++++++++++++++++++++
+ include/linux/phy.h                | 12 ------------
```
Yes, there is a conflict when applying this patch to the openwrt kernel 6.6.90 because the following patch is not present in that version. As a result, there is one fewer line to delete compared to the original patch.

commit 20a4da20e0bd0297bf1ce083362e78b6702f991e 

```diff
--- a/include/linux/phy.h
+++ b/include/linux/phy.h
@@ -1260,6 +1260,7 @@ size_t phy_speeds(unsigned int *speeds, size_t size,
                  unsigned long *mask);
 void of_set_phy_supported(struct phy_device *phydev);
 void of_set_phy_eee_broken(struct phy_device *phydev);
+void of_set_phy_timing_role(struct phy_device *phydev);
 int phy_speed_down_core(struct phy_device *phydev);
 ```

jie-luo89 commented on target/linux/qualcommbe/patches-6.6/910-42-arm64-dts-qcom-Fix-PCS-dts-compile-issue.patch:
```diff
@@ -0,0 +1,36 @@
++				 <&pcs1 UNIPHY_NSS_RX_CLK>,
++				 <&pcs1 UNIPHY_NSS_TX_CLK>,
++				 <&pcs2 UNIPHY_NSS_RX_CLK>,
++				 <&pcs2 UNIPHY_NSS_TX_CLK>,
```
Yes, `pcs<n>` is the expected label format for PCS nodes and is intended to be used when these nodes are merged upstream.

jie-luo89 commented on target/linux/qualcommbe/patches-6.6/901-05-net-phy-qca808x-Add-QCA8084-SerDes-probe-and-remove-.patch:
```diff
@@ -0,0 +1,448 @@
++		if (of_node_name_eq(child, "pcs-phy")) {
++			mdiodev = qca8084_package_pcs_probe(child);
++			if (IS_ERR(mdiodev))
++				return PTR_ERR(mdiodev);
```
Could you please try using the QCA8084 PHY DTS example provided in the DT binding file `Documentation/devicetree/bindings/net/qcom,qca8084.yaml` from this patch series? Also, please remove the `qcom,package-mode` property, as the default package mode is "10g-qxgmii".

quic-rjangir committed:
qualcommbe: Add basic support for IPQ54xx RDP466 and RDP487

Signed-off-by: Ram Chandra Jangir <quic_rjangir@quicinc.com>

quic-rjangir committed:
ipq54xx: Add pinctrl, clk and basic boot support

Signed-off-by: Ram Chandra Jangir <quic_rjangir@quicinc.com>

quic-rjangir committed:
qualcommbe: ipq5424: Add PCIe PHYs controller support

Signed-off-by: Ram Chandra Jangir <quic_rjangir@quicinc.com>

quic-rjangir committed:
qualcommbe: ipq5424: Add basic support for RDP487

Signed-off-by: Ram Chandra Jangir <quic_rjangir@quicinc.com>

hauke commented on target/linux/qualcommbe/image/ipq54xx.mk:
```diff
@@ -0,0 +1,31 @@
+endef
+TARGET_DEVICES += qcom_rdp466
+
+define Device/qcom_rdp487
+	$(call Device/FitImageLzma)
+	KERNEL_LOADADDR := 0x80000000
+	DEVICE_VENDOR := Qualcomm Technologies, Inc.
+	DEVICE_MODEL := RDP487
+	DEVICE_VARIANT := AP-MR03
+	BOARD_NAME := ap-mr02
+	DEVICE_DTS_CONFIG := config@rdp487
+	SOC := ipq5424
+	KERNEL_INSTALL := 1
+	KERNEL_SIZE := 6096k
+	IMAGE_SIZE := 25344k
+	IMAGE/sysupgrade.bin := append-kernel | pad-to 64k | append-rootfs | pad-rootfs | check-size | append-metadata
+endef
+TARGET_DEVICES += qcom_rdp487
```
Please add this in the commit which also adds the dts file for the `qcom_rdp487`.

Commit:
qualcommbe: ipq54xx: Enable RDP487 Ethernet function

RDP487 is connected with QCA8386 switch via PPE port 3 and connected with
QCA8111 PHY via PPE port 2.

Signed-off-by: Luo Jie <jie.luo@oss.qualcomm.com>

quic-rjangir committed:
qualcommbe: ipq95xx: Update the default configs

Disable QCA81XX PHY and IPQ5424 clock controller drivers to avoid asking
user to select during the compilation.

Signed-off-by: Ram Chandra Jangir <rjangir@qti.qualcomm.com>

jie-luo89 commented:
> @quic-rjangir / @mrnuke qca8084 phy usable on ipq95xx platform with this patch set ?
> 
> I have tried to get the qca8084 working via pcs0 using quad / 10g-qxgmii mode. The phy inits fine but I get no link on any of the 8084 attached ports.
> 
> I also saw some stuck clock warnings while booting that seems to only happen when trying to use 8084

QCA8084 is expected to be functional. Please refer to the device tree source file at arch/arm64/boot/dts/qcom/ipq9574-rdp453.dts, which has been updated in this new patch set to enable the QCA8084 PHY in 10G-QXGMII mode.

BuzzBumbleBee commented:
> > @quic-rjangir / @mrnuke qca8084 phy usable on ipq95xx platform with this patch set ?
> > I have tried to get the qca8084 working via pcs0 using quad / 10g-qxgmii mode. The phy inits fine but I get no link on any of the 8084 attached ports.
> > I also saw some stuck clock warnings while booting that seems to only happen when trying to use 8084
> 
> QCA8084 is expected to be functional. Please refer to the device tree source file at arch/arm64/boot/dts/qcom/ipq9574-rdp453.dts, which has been updated in this new patch set to enable the QCA8084 PHY in 10G-QXGMII mode.

Unfortunately it doesn't seem to work for me, ppe and the QCA8084 phy drivers seem to start OK but link will not come up, I have attached my DTS, Kernel logs, mdio scan and ethtool output. There are clock warnings on link up and down commands as well

This device is ipq9574 based with PCS0 used in quad port mode, PCS1 and PSC2 attached to SFP+ modules


DTS
[ipq9574-route10.txt](https://github.com/user-attachments/files/22227200/ipq9574-route10.txt)

ip link set X up warning
[ip_link_warning.txt](https://github.com/user-attachments/files/22227223/ip_link_warning.txt)

mdio scan + ethtool output
[mdio_ethtool.txt](https://github.com/user-attachments/files/22227244/mdio_ethtool.txt)

kernel boot logs
[kernel_logs.txt](https://github.com/user-attachments/files/22227249/kernel_logs.txt)

One thing to note this device was booted tftp un uboot

BuzzBumbleBee commented:
> > @quic-rjangir / @mrnuke qca8084 phy usable on ipq95xx platform with this patch set ?
> > I have tried to get the qca8084 working via pcs0 using quad / 10g-qxgmii mode. The phy inits fine but I get no link on any of the 8084 attached ports.
> > I also saw some stuck clock warnings while booting that seems to only happen when trying to use 8084
> 
> QCA8084 is expected to be functional. Please refer to the device tree source file at arch/arm64/boot/dts/qcom/ipq9574-rdp453.dts, which has been updated in this new patch set to enable the QCA8084 PHY in 10G-QXGMII mode.

@quic-rjangir / @quic-varada 

After some digging this patch set seems to break link up events when SFP+ ports are used, applying the following partial revert resolves thus

```
Index: linux-6.12.43/drivers/net/pcs/pcs-qcom-ipq9574.c
===================================================================
--- linux-6.12.43.orig/drivers/net/pcs/pcs-qcom-ipq9574.c
+++ linux-6.12.43/drivers/net/pcs/pcs-qcom-ipq9574.c
@@ -489,6 +489,112 @@ static int ipq_pcs_config_10gbaser(struc
 	return ipq_pcs_config_mode(qpcs, PHY_INTERFACE_MODE_10GBASER);
 }
 
+static unsigned long ipq_pcs_clock_rate_get_gmii(int speed)
+{
+	unsigned long rate = 0;
+
+	switch (speed) {
+	case SPEED_1000:
+		rate = 125000000;
+		break;
+	case SPEED_100:
+		rate = 25000000;
+		break;
+	case SPEED_10:
+		rate = 2500000;
+		break;
+	default:
+		break;
+	}
+
+	return rate;
+}
+
+static unsigned long ipq_pcs_clock_rate_get_gmiiplus(int speed)
+{
+	unsigned long rate = 0;
+
+	switch (speed) {
+	case SPEED_2500:
+		rate = 312500000;
+		break;
+	default:
+		break;
+	}
+
+	return rate;
+}
+
+static unsigned long ipq_pcs_clock_rate_get_xgmii(int speed)
+{
+	unsigned long rate = 0;
+
+	switch (speed) {
+	case SPEED_10000:
+		rate = 312500000;
+		break;
+	case SPEED_5000:
+		rate = 156250000;
+		break;
+	case SPEED_2500:
+		rate = 78125000;
+		break;
+	case SPEED_1000:
+		rate = 125000000;
+		break;
+	case SPEED_100:
+		rate = 12500000;
+		break;
+	case SPEED_10:
+		rate = 1250000;
+		break;
+	default:
+		break;
+	}
+
+	return rate;
+}
+
+static void
+ipq_pcs_link_up_clock_rate_set(struct ipq_pcs_mii *qunipcs_ch,
+				 phy_interface_t interface,
+				 int speed)
+{
+	struct ipq_pcs *qpcs = qunipcs_ch->qpcs;
+	unsigned long rate = 0;
+
+	switch (interface) {
+	case PHY_INTERFACE_MODE_SGMII:
+	case PHY_INTERFACE_MODE_QSGMII:
+	case PHY_INTERFACE_MODE_PSGMII:
+	case PHY_INTERFACE_MODE_1000BASEX:
+		rate = ipq_pcs_clock_rate_get_gmii(speed);
+		break;
+	case PHY_INTERFACE_MODE_2500BASEX:
+		rate = ipq_pcs_clock_rate_get_gmiiplus(speed);
+		break;
+	case PHY_INTERFACE_MODE_USXGMII:
+	case PHY_INTERFACE_MODE_10GBASER:
+	case PHY_INTERFACE_MODE_10G_QXGMII:
+		rate = ipq_pcs_clock_rate_get_xgmii(speed);
+		break;
+	default:
+		dev_err(qpcs->dev,
+			"interface %s not supported\n", phy_modes(interface));
+		return;
+	}
+
+	if (rate == 0) {
+		dev_err(qpcs->dev, "Invalid PCS clock rate\n");
+		return;
+	}
+
+	clk_set_rate(qunipcs_ch->rx_clk, rate);
+	clk_set_rate(qunipcs_ch->tx_clk, rate);
+
+	fsleep(10000);
+}
+
 static int ipq_pcs_link_up_config_sgmii(struct ipq_pcs *qpcs,
 					int index,
 					unsigned int neg_mode,
@@ -625,9 +731,9 @@ static unsigned int ipq_pcs_inband_caps(
 	case PHY_INTERFACE_MODE_1000BASEX:
 	case PHY_INTERFACE_MODE_USXGMII:
 	case PHY_INTERFACE_MODE_10G_QXGMII:
+        case PHY_INTERFACE_MODE_10GBASER:
 		return LINK_INBAND_DISABLE | LINK_INBAND_ENABLE;
 	case PHY_INTERFACE_MODE_2500BASEX:
-	case PHY_INTERFACE_MODE_10GBASER:
 		return LINK_INBAND_DISABLE;
 	default:
 		return 0;
@@ -750,6 +856,9 @@ static void ipq_pcs_link_up(struct phyli
 	int index = qpcs_mii->index;
 	int ret;
 
+	/* Configure PCS channel interface clock rate */
+	ipq_pcs_link_up_clock_rate_set(qpcs_mii, interface, speed);
+
 	switch (interface) {
 	case PHY_INTERFACE_MODE_SGMII:
 	case PHY_INTERFACE_MODE_QSGMII:
```

jie-luo89 commented:
> > @BuzzBumbleBee, The PCS clock shares the same RCG (clock tree) as the PPE port clock. The clock rate configuration is handled within the PPE MAC link-up function, specifically in ppe_port_mac_speed_clock_rate_set(). Therefore, the PCS driver does not need to adjust the clock rate based on link speed.
> 
> Humm maybe its the `fsleep` or `LINK_INBAND_ENABLE` flag within the link up config ? iv tested again and 100% without the patch I provided the SFP+ ports never get "link up" (keep in mind the SFP+ ports work as expected in current OpenWRT main branch)

OK. Could you please try adding LINK_INBAND_ENABLE for PHY_INTERFACE_MODE_10GBASER, since you have defined managed = "in-band-status"; in your SFP port DTS?

jie-luo89 commented:
> > > > @BuzzBumbleBee, The PCS clock shares the same RCG (clock tree) as the PPE port clock. The clock rate configuration is handled within the PPE MAC link-up function, specifically in ppe_port_mac_speed_clock_rate_set(). Therefore, the PCS driver does not need to adjust the clock rate based on link speed.
> > > 
> > > 
> > > Humm maybe its the `fsleep` or `LINK_INBAND_ENABLE` flag within the link up config ? iv tested again and 100% without the patch I provided the SFP+ ports never get "link up" (keep in mind the SFP+ ports work as expected in current OpenWRT main branch)
> > 
> > 
> > OK. Could you please try adding LINK_INBAND_ENABLE for PHY_INTERFACE_MODE_10GBASER, since you have defined managed = "in-band-status"; in your SFP port DTS?
> 
> Tested and confirmed the only change thats needed is
> 
> ```
> Index: linux-6.12.43/drivers/net/pcs/pcs-qcom-ipq9574.c
> ===================================================================
> --- linux-6.12.43.orig/drivers/net/pcs/pcs-qcom-ipq9574.c
> +++ linux-6.12.43/drivers/net/pcs/pcs-qcom-ipq9574.c
> @@ -624,10 +624,10 @@ static unsigned int ipq_pcs_inband_caps(
>  	case PHY_INTERFACE_MODE_PSGMII:
>  	case PHY_INTERFACE_MODE_1000BASEX:
>  	case PHY_INTERFACE_MODE_USXGMII:
> +	case PHY_INTERFACE_MODE_10GBASER:
>  	case PHY_INTERFACE_MODE_10G_QXGMII:
>  		return LINK_INBAND_DISABLE | LINK_INBAND_ENABLE;
>  	case PHY_INTERFACE_MODE_2500BASEX:
> -	case PHY_INTERFACE_MODE_10GBASER:
>  		return LINK_INBAND_DISABLE;
>  	default:
>  		return 0;
> ```

Cool. Thanks for confirmation.

sidhantgoel commented:
@BuzzBumbleBee That worked friend, thank you very much, also I have another problem, two 10g nics only work if they are initialized in u-boot, so I have patched u-boot to initialize it and then boot into linux, but if it boots into linux without initializing the links dont come up, can you help me there also please.

BuzzBumbleBee commented:
> @BuzzBumbleBee That worked friend, thank you very much, also I have another problem, two 10g nics only work if they are initialized in u-boot, so I have patched u-boot to initialize it and then boot into linux, but if it boots into linux without initializing the links dont come up, can you help me there also please.
> 
> Also, not related to above but do you thing the following warnings about the clocks are imp or can be ignored,
> 

The clock warnings can be ignored for for afaik.

For the 10G NICs not starting, unfortunately I cannot help with that one as I do not have a device with those.

sidhantgoel commented:
@BuzzBumbleBee I am able to bring up the two NICs without u-boot patch with the following kernel patch, but please can you help me with the right way to do this using dts.

```--- a/drivers/net/phy/aquantia/aquantia_main.c
+++ b/drivers/net/phy/aquantia/aquantia_main.c
@@ -215,6 +215,10 @@ static int aqr_config_aneg(struct phy_de
 			      phydev->advertising))
 		reg |= MDIO_AN_VEND_PROV_5000BASET_FULL;
 
+	ret = phy_write_mmd(phydev, MDIO_MMD_PHYXS, 0xc441, 0x08);
+	if (ret < 0)
+		return ret;
+
 	ret = phy_modify_mmd_changed(phydev, MDIO_MMD_AN, MDIO_AN_VEND_PROV,
 				     MDIO_AN_VEND_PROV_1000BASET_HALF |
 				     MDIO_AN_VEND_PROV_1000BASET_FULL |
@@ -281,7 +285,7 @@ static int aqr_config_aneg_set_prot(stru
 
 static int aqr_config_intr(struct phy_device *phydev)
 {
-	bool en = phydev->interrupts == PHY_INTERRUPT_ENABLED;
+	bool en = true;//phydev->interrupts == PHY_INTERRUPT_ENABLED;
 	int err;
 
 	if (en) {
```

quic-vijbal committed:
ipq54xx : Fix DMA engine and MTD nand driver for spi nand probe

Here we are fixing the spi nand device probing for ipq54xx
target with below fixes :
	1. BAM_DESC_CNT_TRSHLD - Avoiding write to this unavailable register in BAM-Lite mode
	   Enabling it to write only in BAM-NDP mode.
	2. Enabling SPI nand support for ipq54xx with bam offset and compatiblity
	3. Update qpic_nand and qpic_bam data in ipq54xx DTS files
	4. Added a new field 'bam_offset' in the NAND properties structure and use it while
	   preparing the command descriptor

Signed-off-by: Vijay Balaji <quic_vijbal@quicinc.com>

quic-vijbal committed:
ipq54xx : Fix squashfs mount using MTD parition name

Here we are fixing squashfs mount with MTD parition name
instead of device number using separate ubi device.

It also enables CONFIG_MTD_UBI_GLUEBI which allows the UBI
subsystem to expose UBI volumes as MTD devices using a "glue" layer

Signed-off-by: Vijay Balaji <quic_vijbal@quicinc.com>

mrnuke commented on target/linux/qualcommbe/patches-6.12/0512-arm64-dts-Add-RDP453-DTS.patch:
```diff
@@ -0,0 +1,467 @@
++                                  "xpcs_rx",
++                                  "xpcs_tx",
++                                  "port_rx",
++                                  "port_tx";
```
The same TX/RX swapping for the resets happens here. Is this intentional?

mrnuke commented on target/linux/qualcommbe/patches-6.12/0512-arm64-dts-Add-RDP453-DTS.patch:
```diff
@@ -0,0 +1,467 @@
++                clocks = <&qca8k_nsscc NSS_CC_SRDS1_SYS_CLK>,
++                         <&qca8k_uniphy1_tx312p5m>,
++                         <&qca8k_uniphy1_rx312p5m>;
++                clock-names = "pcs", "pcs_rx_root", "pcs_tx_root";
```
**pcs_rx_root** is mapped to the TX clock, and vice-versa. Is this intended?

mrnuke reviewed (commented):
I tried to use this series to get the QCA8084 up and running on my ipq9554 based tplink be550. The `lan1` -> `lan3` devices come up. The LEDs on the link partner turn on,  `ethtool` reports a speed and duplex, but no link. I do not see the driver report a link up event and the etherenet interfaces do not pass traffic.

@quic-rjangir, @jie-luo89 , could this be related to the swapping of TX and RX clocks I mention in the inline comments? 

EDIT: My problem was trying to enable `managed = "in-band-status";` on the ports connected to the QCA8084. I removed that, and the link works.
